File: iris-did/src/main/java/city/smartb/iris/did/DIDDocumentBuilder.java
package city.smartb.iris.did;

import city.smartb.iris.did.model.ControledJsonLdObject;
import city.smartb.iris.did.model.DIDAuthentication;
import city.smartb.iris.did.model.DIDService;
import city.smartb.iris.did.model.DIDVerificationMethod;
import city.smartb.iris.jsonld.JsonLdObject;
import city.smartb.iris.jsonld.reader.JsonFieldReader;
import city.smartb.iris.ldproof.LdProof;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class DIDDocumentBuilder extends JsonLdObject {

    public static DIDDocumentBuilder create() {
        return new DIDDocumentBuilder(new LinkedHashMap<>());
    }

    public static DIDDocumentBuilder fromMap(LinkedHashMap<String, Object> json) {
        return new DIDDocumentBuilder(json);
    }

    public DIDDocumentBuilder(Map<String, Object> json) {
        super(json);
    }

    public DIDDocumentBuilder(Map<String, Object> json, JsonFieldReader reader) {
        super(json, reader);
    }

    public DIDDocumentBuilder withId(String id) {
        jsonLdObject.put(JsonLdObject.JSON_LD_ID, id);
        return this;
    }

    public DIDDocumentBuilder withVerificationMethods(List<DIDVerificationMethod> methods) {
        List<Object> methodsJson = methods.parallelStream().map(method -> method.asJson()).collect(Collectors.toList());
        jsonLdObject.put(DIDDocument.JSON_LD_VERIFICATION_METHOD, methodsJson);
        return this;
    }

    public DIDDocumentBuilder withVerificationMethod(DIDVerificationMethod method) {
        List<DIDVerificationMethod> methods = Collections.singletonList(method);
        return this.withVerificationMethods(methods);
    }

    public DIDDocumentBuilder withServices(List<DIDService> services) {
        List<Map<String, Object>> servicesJSon = services.parallelStream().map(service -> service.asJson()).collect(Collectors.toList());
        jsonLdObject.put(DIDDocument.JSON_LD_SERVICE, servicesJSon);
        return this;
    }

    public DIDDocumentBuilder withService(DIDService service) {
        List<DIDService> services = Collections.singletonList(service);
        return this.withServices(services);
    }

    public DIDDocumentBuilder withAuthentications(List<DIDAuthentication> authentications) {
        List<Object> authenticationsJson = authentications.parallelStream().map(authentication -> authentication.toJSON()).collect(Collectors.toList());
        jsonLdObject.put(DIDDocument.JSON_LD_AUTHENTICATION, authenticationsJson);
        return this;
    }

    public DIDDocumentBuilder withAuthentication(DIDAuthentication authentication) {
        List<DIDAuthentication> authentications = Collections.singletonList(authentication);
        return this.withAuthentications(authentications);
    }

    public DIDDocumentBuilder withController(String controller) {
        jsonLdObject.put(ControledJsonLdObject.JSON_LD_CONTROLER, controller);
        return this;
    }

    public List<DIDAuthentication> getProof() {
        return this.get(DIDDocument.JSON_LD_AUTHENTICATION)
                .asListObjects(Object.class)
                .stream()
                .map(it -> new DIDAuthentication(it))
                .collect(Collectors.toList());
    }

    public Map<String, Object> asJson() {
        return this.jsonLdObject;
    }

    public DIDDocument asJson(LdProof proof) {
        this.jsonLdObject.put(LdProof.JSON_LD_PROOF, proof);
        return  new DIDDocument(this.jsonLdObject);
    }
}

File: iris-ld/iris-did/src/main/kotlin/city/smartb/iris/ld/did/DIDDocumentBuilder.kt
package city.smartb.iris.ld.did

import city.smartb.iris.ld.jsonld.JsonLdObject
import city.smartb.iris.ld.jsonld.reader.JsonFieldReader
import city.smartb.iris.ld.ldproof.LdProof
import java.util.stream.Collectors

class DIDDocumentBuilder : JsonLdObject {
    constructor(json: Map<String, Any>) : super(json) {}
    constructor(json: Map<String, Any>, reader: JsonFieldReader) : super(json, reader) {}

    companion object {
        fun create(): DIDDocumentBuilder {
            return DIDDocumentBuilder(LinkedHashMap())
        }

        fun fromMap(json: LinkedHashMap<String, Any>): DIDDocumentBuilder {
            return DIDDocumentBuilder(json)
        }
    }

    fun withId(id: String): DIDDocumentBuilder {
        jsonLdObject[JSON_LD_ID] = id
        return this
    }

    fun withVerificationMethods(methods: List<DIDVerificationMethod>): DIDDocumentBuilder {
        val methodsJson = methods.parallelStream().map { method: DIDVerificationMethod -> method.asJson() }
            .collect(Collectors.toList())
        jsonLdObject[DIDDocument.JSON_LD_VERIFICATION_METHOD] = methodsJson
        return this
    }

    fun withAssertionMethods(items: List<String>): DIDDocumentBuilder {
        jsonLdObject[DIDDocument.JSON_LD_ASSERTION_METHOD] = items
        return this
    }

    fun withCapabilityInvocations(items: List<String>): DIDDocumentBuilder {
        jsonLdObject[DIDDocument.JSON_LD_CAPABILITY_INVOCATION] = items
        return this
    }

    fun withCapabilityDelegations(items: List<String>): DIDDocumentBuilder {
        jsonLdObject[DIDDocument.JSON_LD_CAPABILITY_DELEGATION] = items
        return this
    }

    fun withKeyAgreements(items: List<String>): DIDDocumentBuilder {
        jsonLdObject[DIDDocument.JSON_LD_KEY_AGREEMENT] = items
        return this
    }

    fun withVerificationMethod(method: DIDVerificationMethod): DIDDocumentBuilder {
        val methods = listOf(method)
        return withVerificationMethods(methods)
    }

    fun withServices(services: List<DIDService>): DIDDocumentBuilder {
        val servicesJSon = services.parallelStream().map { service: DIDService -> service.asJson() }
            .collect(Collectors.toList())
        jsonLdObject[DIDDocument.JSON_LD_SERVICE] = servicesJSon
        return this
    }

    fun withService(service: DIDService): DIDDocumentBuilder {
        val services = listOf(service)
        return withServices(services)
    }

    fun withAuthentications(authentications: List<DIDAuthentication>): DIDDocumentBuilder {
        val authenticationsJson =
            authentications.parallelStream().map { authentication: DIDAuthentication -> authentication.toJSON() }
                .collect(Collectors.toList())
        jsonLdObject[DIDDocument.JSON_LD_AUTHENTICATION] = authenticationsJson
        return this
    }

    fun withAuthentication(authentication: DIDAuthentication): DIDDocumentBuilder {
        val authentications = listOf(authentication)
        return withAuthentications(authentications)
    }

    fun withController(controller: String): DIDDocumentBuilder {
        jsonLdObject[ControledJsonLdObject.JSON_LD_CONTROLER] = controller
        return this
    }

    val proof: List<DIDAuthentication>
        get() = this[DIDDocument.JSON_LD_AUTHENTICATION]
            .asListObjects(Any::class.java)
            .stream()
            .map { DIDAuthentication(it)
            }
            .collect(Collectors.toList())

    override fun asJson(): Map<String, Any> {
        return jsonLdObject
    }

    fun asJson(proof: LdProof): DIDDocument {
        jsonLdObject[LdProof.JSON_LD_PROOF] = proof
        return DIDDocument(jsonLdObject)
    }
}
